src/io.jl

```diff
@@ -18,341 +18,93 @@ function ioraise(e)
-# TODO: support other Python interfaces (e.g. TextIO) when possible?
 
 ##########################################################################
 
-# make method tables for text/binary IO objects (called in __io__)
-make_io_methods(text::Bool) =
-    PyMethodDef[
-                PyMethodDef("close", jl_IO_close, METH_NOARGS),
-                PyMethodDef("fileno", jl_IO_fileno, METH_NOARGS),
-                PyMethodDef("flush", jl_IO_flush, METH_NOARGS),
-                PyMethodDef("isatty", jl_IO_isatty, METH_NOARGS),
-                PyMethodDef("readable", jl_IO_readable, METH_NOARGS),
-                PyMethodDef("writable", jl_IO_writable, METH_NOARGS),
-                PyMethodDef("readline", jl_IO_readline, METH_VARARGS),
-                PyMethodDef("readlines", jl_IO_readlines, METH_VARARGS),
-                PyMethodDef("seek", jl_IO_seek, METH_VARARGS),
-                PyMethodDef("seekable", jl_IO_seekable, METH_NOARGS),
-                PyMethodDef("tell", jl_IO_tell, METH_NOARGS),
-                PyMethodDef("writelines", jl_IO_writelines, METH_O),
-                PyMethodDef("read", text ? jl_IO_read_text : jl_IO_read, METH_VARARGS),
-                PyMethodDef("readall", text ? jl_IO_readall_text : jl_IO_readall, METH_NOARGS),
-                PyMethodDef("readinto", jl_IO_readinto, METH_O),
-                PyMethodDef("write", jl_IO_write, METH_O),
-                PyMethodDef() # sentinel
-    ]
-
-
-function pyio_repr(o::PyPtr)
-    o = PyObject(try string("<PyCall.io ",unsafe_pyjlwrap_to_objref(o),">")
-                 catch "<PyCall.io NULL>"; end)
-    oret = o.o
-    o.o = convert(PyPtr, C_NULL) # don't decref
-    return oret
-end
-
 pyio_initialized = false
 function pyio_initialize()
     global pyio_initialized
     if !pyio_initialized::Bool
-        global const jl_IOType =
-            pyjlwrap_type("PyCall.jl_IO",
-                          t -> begin
-                              t.tp_getattro = @pyglobal(:PyObject_GenericGetAttr)
-                              t.tp_methods = pointer(jl_IO_methods)
-                              t.tp_getset = pointer(jl_IO_getset)
-                              t.tp_repr = pyio_repr_ptr
-                          end)
-        global const jl_TextIOType =
-            pyjlwrap_type("PyCall.jl_TextIO",
-                          t -> begin
-                              t.tp_getattro = @pyglobal(:PyObject_GenericGetAttr)
-                              t.tp_methods = pointer(jl_TextIO_methods)
-                              t.tp_getset = pointer(jl_IO_getset)
-                              t.tp_repr = pyio_repr_ptr
-                          end)
+        @eval @pydef type PyIO
+            __init__(self, io::IO; istextio=false) = begin
+                self[:io] = pyjlwrap_new(io) # avoid recursion
+                self[:istextio] = istextio
+                nothing
+            end
+            close(self) = @with_ioraise(close(self[:io]))
```

Comment:
Since we know the type of the `io` member, probably we should do `unsafe_pyjlwrap_to_objref(self["io"])::IO` instead of `self[:io]` to avoid the overhead of the runtime type detection and to add compile-time type information. (Probably via a helper function `pyio_jl(self)` for this.)